File: core/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/UnsafeUtils.scala
package com.github.plokhotnyuk.jsoniter_scala.core

import sun.misc.Unsafe

import scala.util.Try

// FIXME: remove when perf. degradation of String.charAt when iterating through strings will be fixed in JDK 9:
// https://bugs.openjdk.java.net/browse/JDK-8013655
object UnsafeUtils {
  private[this] final val (unsafe, stringValueOffset, stringCoderOffset) = Try {
    val u = {
      val unsafeField = classOf[Unsafe].getDeclaredField("theUnsafe")
      unsafeField.setAccessible(true)
      unsafeField.get(null).asInstanceOf[Unsafe]
    }
    (u,
      u.objectFieldOffset(classOf[String].getDeclaredField("value")),
      u.objectFieldOffset(classOf[String].getDeclaredField("coder")))
  }.getOrElse((null, 0L, 0L))

  private[jsoniter_scala] final def getLatin1Array(s: String): Array[Byte] =
    if (stringCoderOffset == 0 || unsafe.getByte(s, stringCoderOffset) != 0) null
    else unsafe.getObject(s, stringValueOffset).asInstanceOf[Array[Byte]]
}

File: core/src/main/java/com/github/plokhotnyuk/jsoniter_scala/core/UnsafeUtils.java
package com.github.plokhotnyuk.jsoniter_scala.core;

import java.lang.reflect.Field;

// FIXME: remove when perf. degradation of String.charAt when iterating through strings will be fixed in JDK 9+:
// https://bugs.openjdk.java.net/browse/JDK-8013655
public class UnsafeUtils {
    private static final sun.misc.Unsafe UNSAFE;
    private static final long STRING_CODER_OFFSET;
    private static final long STRING_VALUE_OFFSET;

    static {
        sun.misc.Unsafe u = null;
        long sco = 0, svo = 0;
        try {
            Field f = sun.misc.Unsafe.class.getDeclaredField("theUnsafe");
            f.setAccessible(true);
            u = (sun.misc.Unsafe) f.get(null);
            sco = u.objectFieldOffset(String.class.getDeclaredField("coder"));
            svo = u.objectFieldOffset(String.class.getDeclaredField("value"));
        } catch (Throwable e) {
            // ignore
        }
        UNSAFE = u;
        STRING_CODER_OFFSET = sco;
        STRING_VALUE_OFFSET = svo;
    }

    static byte[] getLatin1Array(String s) {
        if (STRING_CODER_OFFSET == 0 || s == null || UNSAFE.getByte(s, STRING_CODER_OFFSET) != 0) {
            return null;
        }
        return (byte[]) UNSAFE.getObject(s, STRING_VALUE_OFFSET);
    }
}
